internal/step/foreach/provider.go

```diff
@@ -368,6 +370,11 @@ type runningStep struct {
 
 func (r *runningStep) ProvideStageInput(stage string, input map[string]any) error {
 	r.lock.Lock()
+	if r.closed {
+		r.logger.Debugf("exiting foreach ProvideStageInput due to step being closed")
+		r.lock.Unlock()
```

Comment:
I kinda wonder if deferring an unlock wouldn't be cleaner than unlocking here and in each of the cases below ... it'd a really easy for this logic to get broken.